README.md
# Codam Header

![Build](https://github.com/skrtks/clion_codam_header/workflows/Build/badge.svg)    

<!-- Plugin description -->
Welcome to the Codam Header plugin for CLion. Add header to .c and .h files via <kbd>tools</kbd> > <kbd>Add codam header</kbd> or cmd-opt-h.
<!-- Plugin description end -->

## Installation

- Using IDE built-in plugin system:
  
  <kbd>Preferences</kbd> > <kbd>Plugins</kbd> > <kbd>Marketplace</kbd> > <kbd>Search for "clion_codam_header"</kbd> >
  <kbd>Install Plugin</kbd>
  
- Manually:

  Download the [latest release](https://github.com/skrtks/clion_codam_header/releases/latest) and install it manually using
  <kbd>Preferences</kbd> > <kbd>Plugins</kbd> > <kbd>⚙️</kbd> > <kbd>Install plugin from disk...</kbd>


---
Plugin based on the [IntelliJ Platform Plugin Template][template].

[template]: https://github.com/JetBrains/intellij-platform-plugin-template


CHANGELOG.md
<!-- Keep a Changelog guide -> https://keepachangelog.com -->

# clion_codam_header Changelog

## [Unreleased]
### Added
- Now only targeting CLion instead of all JetBrains IDE's


src/main/java/Save.java
//import com.intellij.openapi.command.WriteCommandAction;
//import com.intellij.openapi.components.ApplicationComponent;
//import com.intellij.openapi.fileEditor.FileDocumentManager;
//import com.intellij.openapi.project.ProjectManager;
//import com.intellij.openapi.util.TextRange;
//import com.intellij.openapi.vfs.VirtualFile;
//import com.intellij.openapi.vfs.VirtualFileEvent;
//import com.intellij.openapi.vfs.VirtualFileListener;
//import com.intellij.openapi.vfs.VirtualFileManager;
//import org.jetbrains.annotations.NotNull;
//
//import java.text.DateFormat;
//import java.text.SimpleDateFormat;
//import java.util.Date;
//import java.util.Objects;
//
//public class Save implements ApplicationComponent {
//    public Save() {
//    }
//
//    @Override
//    public void initComponent() {
//        VirtualFileManager.getInstance().addVirtualFileListener(new VirtualFileListener() {
//            public void contentsChanged(@NotNull VirtualFileEvent event) {
//                VirtualFile file = event.getFile();
//                String filename = file.getName();
//                String extension = file.getExtension();
//                String start = Objects.requireNonNull(FileDocumentManager.getInstance().getDocument(file)).getText(new TextRange(0, 5));
//                if (!start.equals("/* **") && !start.equals("# ***"))
//                    return;
//                DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
//                Date date = new Date();
//                while (filename.length() < 51)
//                    filename += ' ';
//                String user = "by " + System.getenv("USER");
//                while (user.length() < 17)
//                    user += ' ';
//                String header;
//                if (filename.contains("Makefile"))
//                    header = "#    Updated: " + dateFormat.format(date) + " " + user + "########   odam.nl          #";
//                else
//                    header = "/*   Updated: " + dateFormat.format(date) + " " + user + "########   odam.nl         */";
//
//                Runnable runnable = () -> Objects.requireNonNull(FileDocumentManager.getInstance().getDocument(file)).replaceString(648, 648 + header.length(), header);
//                WriteCommandAction.runWriteCommandAction(ProjectManager.getInstance().getOpenProjects()[0], runnable);
//            }
//        });
//    }
//
//    @Override
//    public void disposeComponent() {
//    }
//
//    @Override
//    @NotNull
//    public String getComponentName() {
//        return "Save";
//    }
//}

src/main/java/Generate.java
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.command.WriteCommandAction;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Objects;

import static com.intellij.openapi.actionSystem.CommonDataKeys.VIRTUAL_FILE;

public class Generate extends AnAction
{
    @Override
    public void actionPerformed(AnActionEvent AnActionEvent)
    {
        DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
        Date date = new Date();
        VirtualFile file = AnActionEvent.getData(VIRTUAL_FILE);
        assert file != null;
        String filename = file.getName();
        while (filename.length() < 51)
            filename += ' ';
        String user = "By: " + System.getenv("USER") + " " + "<" + System.getenv("USER") + "@student.codam.nl>";
        while (user.length() < 47)
            user += ' ';
        String user2 = "by " + System.getenv("USER");
        while (user2.length() < 17)
            user2 += ' ';
        String user3 = "by " + System.getenv("USER");
        while (user3.length() < 17)
            user3 += ' ';
        String startComment = "/*";
        String endComment = "*/";
        if (filename.contains("Makefile"))
        {
            startComment = "#";
            endComment = "#";
        }
        String header = startComment + " " + (startComment.length() == 1 ? "*" : "") + "**************************************************************************" + (endComment.length() == 1 ? "*" : "") + " " + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "                                                                            " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "                                                        ::::::::            " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "   " + filename + ":+:    :+:            " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "                                                     +:+                    " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "   " + user + "  +#+                     " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "                                                   +#+                      " + (endComment.length() == 1 ? ' ': "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "   Created: " + dateFormat.format(date) + " " + user2 + "#+#    #+#                 " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? " " : "") + "   Updated: " + dateFormat.format(date) + " " + user3 + "########   odam.nl         " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + (startComment.length() == 1 ? ' ' : "") + "                                                                            " + (endComment.length() == 1 ? " ": "") + endComment + "\n" +
                startComment + " " + (startComment.length() == 1 ? "*" : "") + "**************************************************************************" + (endComment.length() == 1 ? "*" : "") + " " + endComment + "\n\n";

        Runnable runnable = () -> Objects.requireNonNull(AnActionEvent.getData(LangDataKeys.EDITOR)).getDocument().insertString(0, header);
        WriteCommandAction.runWriteCommandAction(getEventProject(AnActionEvent), runnable);
    }
}
